ocaml/xapi/xapi_vm_helpers.ml

```diff
@@ -410,18 +410,24 @@ let assert_usbs_available ~__context ~self ~host =
       ]))
     )
 
-(* Get SR-IOV Vifs by return a list of [(network1,(required_num1,PCI1));(network2,(required_num2,PCI2))....] 
+(* 1.To avoid redundant checks,if VF had been reserved, then it's no need to check the ability again.
+   2.Get SR-IOV Vifs by return a list of [(network1,(required_num1,PCI1));(network2,(required_num2,PCI2))....]
    Raise exn immediately when found Idle VF nums < required_num *)
 let assert_netsriov_available ~__context ~self ~host =
   let sriov_networks = List.fold_left (fun acc vif ->
-      let network = Db.VIF.get_network ~__context ~self:vif in
-      try
-        let required, pif = List.assoc network acc in
-        (network,(required + 1,pif)) :: (List.remove_assoc network acc)
-      with Not_found ->
-        match Xapi_network_sriov_helpers.get_local_underlying_pif ~__context ~network ~host with
-        | Some pif -> (network,(1,pif)) :: acc
-        | None -> acc
+      let reserved_pci = Db.VIF.get_reserved_pci ~__context ~self:vif in
```

Comment:
How about to move the check into SR-IOV logic?
We may need to make sure it is a SR-IOV VF firstly, then check if it has been reserved.

Reply:
Yes, but seems that we determine a VIF is SR-IOV or not is to whether we can get the underlying PIF.
While `get_local_underlying_pif` may consumes much effort I think, (it will call `get_local_underlying_pif` in the end which is recursive function).
Since plain VIF won't set the `reserved_pci` at all, if it's not `Ref.null` then must be SR-IOV VIF which has been reserved already, then we do not need to `get_local_underlying_pif` below. Right?
Correct me if it's wrong, thanks for comment.

Reply:
At the moment, it is correct. I just have a little bit concern on treating `reserved_pci` standalone as an indicator of a SR-IOV VF-backed VIF.